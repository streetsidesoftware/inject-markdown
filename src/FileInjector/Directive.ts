import { parseRelativeUrl, type RelURL } from '../util/url_helper.js';

const injectDirectiveRegExp = /^[ \t]*<!--+\s*@@inject(?<type>|-start|-end|-code)[:\s]\s*(?<file>.*?)-+->$/;

export type DirectiveType = 'start' | 'end' | 'code';

export interface Directive {
    type: DirectiveType;
    file: RelURL | undefined;
}

export function parseDirective(html: string): Directive | undefined {
    const m = html.match(injectDirectiveRegExp);
    if (!m || !m.groups) return undefined;

    const filePath = m.groups['file'].trim();
    const file = (filePath && parseRelativeUrl(filePath)) || undefined;
    const isEnd = m.groups['type'] === '-end';
    const isCode = (!isEnd && !file?.pathname.toLowerCase().endsWith('.md')) || m.groups['type'] === '-code';

    const d: Directive = {
        type: isEnd ? 'end' : isCode ? 'code' : 'start',
        file,
    };
    return d;
}
