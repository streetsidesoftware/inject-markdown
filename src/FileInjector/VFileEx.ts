import type { Data as VFileData, MessageOptions as VFileMessageOptions } from 'vfile';
import { VFile } from 'vfile';

import type { BufferEncoding } from '../FileSystemAdapter/FileSystemAdapter.js';

export type MessageOptions = VFileMessageOptions['place'];

export interface FileData extends VFileData {
    encoding: BufferEncoding;
    fileUrl: URL;
    cwdUrl?: URL;
    hasInjections?: boolean;
}

export class VFileEx extends VFile {
    readonly fileUrl: URL;
    data: FileData;

    constructor(value: string, data: FileData) {
        super({ path: data.fileUrl.pathname, value, data });
        this.fileUrl = data.fileUrl;
        this.data = data;
    }

    error(reason: string, place?: MessageOptions): ReturnType<VFile['message']> {
        const msg = this.message(reason, place);
        msg.fatal = true;
        return msg;
    }
}

export function isVFileEx(file: VFile | VFileEx): file is VFileEx {
    return file instanceof VFileEx;
}
