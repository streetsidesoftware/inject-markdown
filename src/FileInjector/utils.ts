import type { BufferEncoding } from '../FileSystemAdapter/FileSystemAdapter.js';

export function toString(content: string | Buffer | Uint8Array, encoding: BufferEncoding): string {
    if (typeof content === 'string') return content;

    if (content instanceof Buffer) {
        return content.toString(encoding);
    }

    return Buffer.from(content).toString(encoding);
}

export function toError(e: unknown): Error {
    if (e && typeof e === 'object') return e as Error;
    if (typeof e === 'string') return new Error(e);
    return new Error('Unknown');
}
