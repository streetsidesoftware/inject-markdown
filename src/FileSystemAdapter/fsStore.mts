import { pathToFileURL } from 'url';

import { PathLike } from './FileSystemAdapter.js';

export interface Store<T> {
    get(file: PathLike): T | undefined;
    set(file: PathLike, data: T | undefined): void;
}

export function createStore<T>(): Store<T> {
    const store = new Map<string, T | undefined>();

    function get(file: PathLike) {
        const p = normalizePath(file).toString();
        return store.get(p);
    }

    function set(file: PathLike, data: T | undefined) {
        const p = normalizePath(file).href;
        if (typeof data === undefined) {
            store.delete(p);
            return;
        }
        store.set(p, data);
    }

    return { get, set };
}

export function normalizePath(pathLike: PathLike): URL {
    pathLike = typeof pathLike !== 'string' ? pathLike.href : pathLike;
    const url = pathToFileURL(pathLike);
    url.hash = '';
    return url;
}
