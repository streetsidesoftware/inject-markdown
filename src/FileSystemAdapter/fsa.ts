import * as fs from 'fs/promises';
import { BufferEncoding, FileSystemAdapter, PathLike } from './FileSystemAdapter.js';
import { isURL } from '../util/url_helper.js';
import fetch from 'node-fetch';

async function readFile(file: PathLike, encoding: BufferEncoding): Promise<string> {
    if (isURL(file)) {
        if (file.protocol === 'file:') {
            return await fs.readFile(file, encoding);
        }
        return await fetchUrl(file);
    }
    return await fs.readFile(file, encoding);
}

export function nodeFsa(): FileSystemAdapter {
    const fsa: FileSystemAdapter = {
        readFile,
        mkdir: fs.mkdir,
        writeFile: fs.writeFile,
    };

    return fsa;
}

async function fetchUrl(url: URL): Promise<string> {
    const response = await fetch(mapUrl(url));
    return await response.text();
}

function mapUrl(url: URL): string {
    let href = url.href;

    // Remap GitHub blobs
    href = href.replace(/^https?:\/\/github.com\/([^/]+\/[^/]+)\/blob\//, 'https://raw.githubusercontent.com/$1/');

    return href;
}
