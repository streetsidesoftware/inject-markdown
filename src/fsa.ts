import * as fs from 'fs/promises';
import { BufferEncoding, FileSystemAdapter, PathLike } from './FileSystemAdapter.js';
import { isURL } from './url_helper.js';
import fetch from 'node-fetch';

async function readFile(file: PathLike, encoding: BufferEncoding): Promise<string> {
    if (isURL(file)) {
        if (file.protocol === 'file:') {
            return await fs.readFile(file, encoding);
        }
        const response = await fetch(file.href);
        return await response.text();
    }
    return await fs.readFile(file, encoding);
}

export function nodeFsa(): FileSystemAdapter {
    const fsa: FileSystemAdapter = {
        readFile,
        mkdir: fs.mkdir,
        writeFile: fs.writeFile,
    };

    return fsa;
}
