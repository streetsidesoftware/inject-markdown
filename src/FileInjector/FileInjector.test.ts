import { fileURLToPath, pathToFileURL } from 'node:url';

import * as path from 'path';
import { format } from 'util';
import { describe, expect, MockedFunction, test, vi } from 'vitest';

import type { BufferEncoding, FileSystemAdapter, PathLike } from '../FileSystemAdapter/FileSystemAdapter.js';
import { nodeFsa } from '../FileSystemAdapter/fsa.js';
import { createStore, normalizePath, type Store } from '../FileSystemAdapter/fsStore.mjs';
import { relativePath } from '../util/url_helper.js';
import { FileInjector, Logger } from './FileInjector.js';

const __file__ = fileURLToPath(import.meta.url);
const __dirname__ = path.dirname(__file__);
const __root__ = path.join(__dirname__, '../..');

const appFsa = nodeFsa();

const oc = (e: unknown) => expect.objectContaining(e);

describe('FileInjector', () => {
    test('FileInjector', () => {
        const fi = new FileInjector(createFSA(), {});
        expect(fi).toBeDefined();
    });

    test.each`
        file           | options                              | expectedResult
        ${'README.md'} | ${{ cwd: 'fixtures/no-injections' }} | ${oc({ hasChanged: false })}
    `('processFile no change', async ({ file, options, expectedResult }) => {
        options.cwd = options.cwd || __root__;
        options.logger = createLogger();
        const fsa = createFSA();
        const fi = new FileInjector(fsa, options);
        const r = await fi.processFile(file);
        expect(r).toEqual(expectedResult);
        expect(fsa.mkdir).not.toBeCalled();
    });

    test.each`
        file                                 | options                                                  | expectedResult               | expectedFile
        ${'fixtures/vacations/vacations.md'} | ${{}}                                                    | ${oc({ hasChanged: true })}  | ${'fixtures/vacations/vacations.md'}
        ${'fixtures/vacations/vacations.md'} | ${{ outputDir: '_out_' }}                                | ${oc({ hasChanged: true })}  | ${'_out_/fixtures/vacations/vacations.md'}
        ${'fixtures/vacations/vacations.md'} | ${{ outputDir: '_out_', verbose: true }}                 | ${oc({ hasChanged: true })}  | ${'_out_/fixtures/vacations/vacations.md'}
        ${'fixtures/vacations/vacations.md'} | ${{ outputDir: '_out_', silent: true }}                  | ${oc({ hasChanged: true })}  | ${'_out_/fixtures/vacations/vacations.md'}
        ${'vacations.md'}                    | ${{ cwd: 'fixtures/vacations/', outputDir: '_out_' }}    | ${oc({ hasChanged: true })}  | ${'_out_/vacations.md'}
        ${'README.md'}                       | ${{ cwd: 'fixtures/no-injections', outputDir: '_out_' }} | ${oc({ hasChanged: false })} | ${'_out_/README.md'}
        ${'fixtures/code/README.md'}         | ${{}}                                                    | ${oc({ hasChanged: true })}  | ${'fixtures/code/README.md'}
    `('processFile $file $options', async ({ file, options, expectedResult, expectedFile }) => {
        const logger = createLogger();
        options.cwd = options.cwd || __root__;
        options.color = options.color ?? false;
        options.logger = logger;
        expectedFile = path.resolve(__root__, expectedFile);
        const fsa = createFSA();
        const fi = new FileInjector(fsa, options);
        const r = await fi.processFile(file);
        expect(r).toEqual(expectedResult);
        expect(fsa.mkdir).toHaveBeenCalledWith(path.dirname(expectedFile), { recursive: true });
        expect(normalizeWriteFileCalls(fsa.writeFile)).toMatchSnapshot();
        expect(logger.history).toMatchSnapshot();
    });

    test.each`
        file                                 | options                                 | expectedResult
        ${'fixtures/vacations/vacations.md'} | ${{ outputDir: '_out_' }}               | ${oc({ hasChanged: true, skipped: false, written: true })}
        ${'fixtures/vacations/vacations.md'} | ${{ outputDir: '_out_', dryRun: true }} | ${oc({ hasChanged: true, skipped: true, written: false })}
    `('processFile($file, $options)', async ({ file, options, expectedResult }) => {
        const logger = createLogger();
        options.cwd = options.cwd || __root__;
        options.color = options.color ?? false;
        options.logger = logger;
        const fsa = createFSA();
        const fi = new FileInjector(fsa, options);
        const r = await fi.processFile(file);
        expect(r).toEqual(expectedResult);
        expect(normalizeWriteFileCalls(fsa.writeFile)).toMatchSnapshot();
        expect(logger.history).toMatchSnapshot();
    });
});

function normalizeWriteFileCalls(
    writeFile: MockedFileSystemAdapter['writeFile']
): MockedFileSystemAdapter['writeFile']['mock']['calls'] {
    const calls = writeFile.mock.calls;
    const cwd = pathToFileURL('.');
    const normalized = calls.map(
        ([pathLike, data, encoding]) =>
            [relativePath(cwd, normalizePath(pathLike)).toString(), data, encoding] as Parameters<
                FileSystemAdapter['writeFile']
            >
    );
    return normalized;
}

type MockedFileSystemAdapter<T extends FileSystemAdapter = FileSystemAdapter> = {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [K in keyof T]: T[K] extends (...args: any) => any ? MockedFunction<T[K]> : T[K];
};
interface FSA extends MockedFileSystemAdapter {
    store: Store<string>;
}

function createFSA(): FSA {
    const store = createStore<string>();

    async function readFile(p: PathLike, e: BufferEncoding): Promise<string> {
        const found = store.get(p);
        if (typeof found === 'string') return found;
        const data = await appFsa.readFile(p, e);
        store.set(p, data);
        return data;
    }

    async function writeFile(file: PathLike, data: string, _encoding: string) {
        store.set(file, data);
    }

    const fsaMethods: FileSystemAdapter = {
        readFile: vi.fn().mockImplementation(readFile),
        writeFile: vi.fn().mockImplementation(writeFile),
        mkdir: vi.fn().mockImplementation(async (_path: PathLike) => undefined),
    };

    const ma = fsaMethods as MockedFileSystemAdapter;

    const fsa: FSA = {
        ...ma,
        store,
    };

    return fsa;
}

function createLogger() {
    type Target = 'error' | 'log' | 'warn' | 'stdout' | 'stderr';
    const history: { target: Target; text: string }[] = [];

    function f(target: Target): typeof console.log {
        function log(...params: Parameters<typeof console.log>) {
            history.push({ target, text: format(...params) });
        }
        return log;
    }
    const log = f('log');
    const error = f('error');
    const warn = f('warn');
    const stderr = f('stderr');
    const stdout = f('stdout');

    const logger = {
        log: vi.fn().mockImplementation(log),
        error: vi.fn(error),
        warn: vi.fn(warn),
        writeStdout: vi.fn((t) => stdout(t)),
        writeStderr: vi.fn((t) => stderr(t)),
        history,
    };
    return logger satisfies Logger;
}
