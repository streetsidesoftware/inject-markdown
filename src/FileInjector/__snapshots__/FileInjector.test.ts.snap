// Vitest Snapshot v1

exports[`FileInjector > processFile 'README.md' { cwd: 'fixtures/no-injections', …(1) } 1`] = `
[
  [
    "_out_/README.md",
    "# Without injections

This file does not have any injections.
",
    "utf8",
  ],
]
`;

exports[`FileInjector > processFile 'README.md' { cwd: 'fixtures/no-injections', …(1) } 2`] = `
[
  {
    "target": "stderr",
    "text": "README.md ...",
  },
  {
    "target": "stderr",
    "text": " done.
",
  },
]
`;

exports[`FileInjector > processFile 'fixtures/code/README.md' {} 1`] = `
[
  [
    "fixtures/code/README.md",
    "# This is a sample README

## Import Code

It contains some code:

<!--- @@inject: code.ts --->

\`\`\`ts
export function sayHello(name: string): string {
    return \`Hello \${name}\`;
}
\`\`\`

<!--- @@inject-end: code.ts --->

## Import \`json\` as \`jsonc\`

<!--- @@inject-code: sample.json#lang=jsonc --->

\`\`\`jsonc
{
    \\"name\\": \\"Sample\\"
}
\`\`\`

<!--- @@inject-end: sample.json#lang=jsonc --->

## Import Markdown as Code

It is also possible to inject markdown:

<!--- @@inject-code: example.md --->

\`\`\`markdown
# Example

This is an example bit of markdown.

- first
- second
- third
\`\`\`

<!--- @@inject-end: example.md --->

## Import a section from a Markdown file

<!--- @@inject: chapters.md#Chapter 3: Directives --->

## Chapter 3: Directives

-   \`@@inject: <markdown_file.md>[#heading]\` and \`@@inject-start:  <markdown_file.md>[#heading]\` -- injects the contents of a markdown file.
    -   \`<markdown_file.md>\` -- the file to import
    -   \`heading\` -- optional heading to extract.
-   \`@@inject: <non-markdown-file>[#lang]\`, \`@@inject-start:  <non-markdown-file>[#lang]\`, and \`@@inject-code: <file>[#lang]\`
    -   \`<non-markdown-file>\`, \`<file>\` -- the file to import
    -   \`lang\` -- optional language to use for the code bock.

<!--- @@inject-end: chapters.md#Chapter 3: Directives --->

## Import from lines from GitHub

<img width=\\"711\\" alt=\\"image\\" src=\\"https://user-images.githubusercontent.com/3740137/210188786-28704fe3-cc2f-447c-97fc-d27715dabbdc.png\\">

\`\`\`
<!--- @@inject: https://github.com/streetsidesoftware/inject-markdown/blob/d7de2f5fe/src/app.mts#L15-L19 --->
\`\`\`

<!--- @@inject: https://github.com/streetsidesoftware/inject-markdown/blob/d7de2f5fe/src/app.mts#L15-L19 --->

\`\`\`typescript
async function version(): Promise<string> {
    const pathSelf = fileURLToPath(import.meta.url);
    const pathPackageJson = path.join(path.dirname(pathSelf), '../package.json');
    const packageJson = JSON.parse(await fs.readFile(pathPackageJson, 'utf8'));
    return (typeof packageJson === 'object' && packageJson?.version) || '0.0.0';
\`\`\`

<!--- @@inject-end: https://github.com/streetsidesoftware/inject-markdown/blob/d7de2f5fe/src/app.mts#L15-L19 --->

## GitHub Footnote

|        | version | Node | Support               | End-Of-Life |
| :----- | :------ | :--- | :-------------------- | :---------- |
| cspell | 6.x     | 14.x | In Active Development | TBD         |
| cspell | 5.x     | 12.x | Paid support only[^1] | 2022-10-01  |
| cspell | 4.x     | 10.x | Paid support only[^1] | 2022-05-01  |

[^1]: [Support - Street Side Software](https://streetsidesoftware.com/support/#maintenance-agreements)
",
    "utf8",
  ],
]
`;

exports[`FileInjector > processFile 'fixtures/code/README.md' {} 2`] = `
[
  {
    "target": "stderr",
    "text": "fixtures/code/README.md ...",
  },
  {
    "target": "stderr",
    "text": " done.
",
  },
]
`;

exports[`FileInjector > processFile 'fixtures/headers/README.md' {} 1`] = `
[
  [
    "fixtures/headers/README.md",
    "# Injecting Headers

<!--- @@inject: content.md#Nested --->

### Nested

This is nested under Section 1

<!--- @@inject-end: content.md#Nested --->

<!--- @@inject: content.md#heading=Nested&L11-L99 --->

### Nested

This is nested under Section 3

<!--- @@inject-end: content.md#heading=Nested&L11-L99 --->

<!--- @@inject: content.md#Nested&L11-L99 --->

### Nested

This is nested under Section 3

<!--- @@inject-end: content.md#Nested&L11-L99 --->

<!--- @@inject: content.md#Missing --->

<!--- header: \\"Missing\\" not found.  --->

<!--- @@inject-end: content.md#Missing --->
",
    "utf8",
  ],
]
`;

exports[`FileInjector > processFile 'fixtures/headers/README.md' {} 2`] = `
[
  {
    "target": "stderr",
    "text": "fixtures/headers/README.md ...",
  },
  {
    "target": "stderr",
    "text": " done.
",
  },
]
`;

exports[`FileInjector > processFile 'fixtures/quotes/README.md' {} 1`] = `
[
  [
    "fixtures/quotes/README.md",
    "# Block Quotes

It is possible to Block Quote injected content.

\`\`\`markdown
<!--- @@inject: include.md#heading=Section 2&quote --->
\`\`\`

<!--- @@inject: include.md#Section 2&quote --->

> ## Section 2
>
> This is section 2 and its content.
>
> -   One
> -   Two
> -   Three
>
> \`\`\`json
> { \\"debug\\": true }
> \`\`\`

<!--- @@inject-end: include.md#Section 2&quote --->

<!--- @@inject: include.md#Section 2&quote&code --->

> \`\`\`\`markdown
> ## Section 2
>
> This is section 2 and its content.
>
> *   One
> *   Two
> *   Three
>
> \`\`\`json
> { \\"debug\\": true }
> \`\`\`
> \`\`\`\`

<!--- @@inject-end: include.md#Section 2&quote&code --->

<!--- @@inject: include.md#L13-L18&quote --->

> This is section 2 and its content.
>
> -   One
> -   Two
> -   Three

<!--- @@inject-end: include.md#L13-L18&quote --->

<!--- @@inject: include.md#heading=Nested&L22-L31&quote --->

> ### Nested
>
> This is nested under Section 3

<!--- @@inject-end: include.md#heading=Nested&L22-L31&quote --->
",
    "utf8",
  ],
]
`;

exports[`FileInjector > processFile 'fixtures/quotes/README.md' {} 2`] = `
[
  {
    "target": "stderr",
    "text": "fixtures/quotes/README.md ...",
  },
  {
    "target": "stderr",
    "text": " done.
",
  },
]
`;

exports[`FileInjector > processFile 'fixtures/vacations/vacations.md' { outputDir: '_out_' } 1`] = `
[
  [
    "_out_/fixtures/vacations/vacations.md",
    "# Vacations

## Destinations

<!--- @@inject: parts/destinations.md --->

## Amsterdam

Amsterdam is full of beautiful canals.

## Orlando

There are many amusement parks in Orlando.

## New York City

It is a wonderful place to catch a show.

## San Francisco

Do not miss out on the Golden Gate Bridge.

<!--- @@inject-end: parts/destinations.md --->

-   Prices
    <!--- @@inject: parts/prices.md --->
    ## Data
    | Name             | City      | Cost   |
    | ---------------- | --------- | ------ |
    | Parade of Lights | Amsterdam | €100/p |
    | Disney World     | Orlando   | $200/p |
    <!--- @@inject-end: parts/prices.md --->

# Highlight Destination (Amsterdam)

<!--- @@inject: parts/destinations.md#Amsterdam --->

## Amsterdam

Amsterdam is full of beautiful canals.

<!--- @@inject-end: parts/destinations.md#Amsterdam --->

## Code

<!--- @@inject-start: parts/code.md --->

## Example

\`\`\`ts
import * as fs from 'fs/promises';
import * as path from 'path';
\`\`\`

<!--- @@inject-end: parts/code.md --->

<pre>
<!--- @@inject: parts/destinations.md -->
</pre>

\`\`\`markdown
<!--- @@inject: parts/destinations.md -->
\`\`\`
",
    "utf8",
  ],
]
`;

exports[`FileInjector > processFile 'fixtures/vacations/vacations.md' { outputDir: '_out_' } 2`] = `
[
  {
    "target": "stderr",
    "text": "fixtures/vacations/vacations.md ...",
  },
  {
    "target": "stderr",
    "text": " done.
",
  },
]
`;

exports[`FileInjector > processFile 'fixtures/vacations/vacations.md' { outputDir: '_out_', silent: true } 1`] = `
[
  [
    "_out_/fixtures/vacations/vacations.md",
    "# Vacations

## Destinations

<!--- @@inject: parts/destinations.md --->

## Amsterdam

Amsterdam is full of beautiful canals.

## Orlando

There are many amusement parks in Orlando.

## New York City

It is a wonderful place to catch a show.

## San Francisco

Do not miss out on the Golden Gate Bridge.

<!--- @@inject-end: parts/destinations.md --->

-   Prices
    <!--- @@inject: parts/prices.md --->
    ## Data
    | Name             | City      | Cost   |
    | ---------------- | --------- | ------ |
    | Parade of Lights | Amsterdam | €100/p |
    | Disney World     | Orlando   | $200/p |
    <!--- @@inject-end: parts/prices.md --->

# Highlight Destination (Amsterdam)

<!--- @@inject: parts/destinations.md#Amsterdam --->

## Amsterdam

Amsterdam is full of beautiful canals.

<!--- @@inject-end: parts/destinations.md#Amsterdam --->

## Code

<!--- @@inject-start: parts/code.md --->

## Example

\`\`\`ts
import * as fs from 'fs/promises';
import * as path from 'path';
\`\`\`

<!--- @@inject-end: parts/code.md --->

<pre>
<!--- @@inject: parts/destinations.md -->
</pre>

\`\`\`markdown
<!--- @@inject: parts/destinations.md -->
\`\`\`
",
    "utf8",
  ],
]
`;

exports[`FileInjector > processFile 'fixtures/vacations/vacations.md' { outputDir: '_out_', silent: true } 2`] = `[]`;

exports[`FileInjector > processFile 'fixtures/vacations/vacations.md' { outputDir: '_out_', verbose: true } 1`] = `
[
  [
    "_out_/fixtures/vacations/vacations.md",
    "# Vacations

## Destinations

<!--- @@inject: parts/destinations.md --->

## Amsterdam

Amsterdam is full of beautiful canals.

## Orlando

There are many amusement parks in Orlando.

## New York City

It is a wonderful place to catch a show.

## San Francisco

Do not miss out on the Golden Gate Bridge.

<!--- @@inject-end: parts/destinations.md --->

-   Prices
    <!--- @@inject: parts/prices.md --->
    ## Data
    | Name             | City      | Cost   |
    | ---------------- | --------- | ------ |
    | Parade of Lights | Amsterdam | €100/p |
    | Disney World     | Orlando   | $200/p |
    <!--- @@inject-end: parts/prices.md --->

# Highlight Destination (Amsterdam)

<!--- @@inject: parts/destinations.md#Amsterdam --->

## Amsterdam

Amsterdam is full of beautiful canals.

<!--- @@inject-end: parts/destinations.md#Amsterdam --->

## Code

<!--- @@inject-start: parts/code.md --->

## Example

\`\`\`ts
import * as fs from 'fs/promises';
import * as path from 'path';
\`\`\`

<!--- @@inject-end: parts/code.md --->

<pre>
<!--- @@inject: parts/destinations.md -->
</pre>

\`\`\`markdown
<!--- @@inject: parts/destinations.md -->
\`\`\`
",
    "utf8",
  ],
]
`;

exports[`FileInjector > processFile 'fixtures/vacations/vacations.md' { outputDir: '_out_', verbose: true } 2`] = `
[
  {
    "target": "stderr",
    "text": "fixtures/vacations/vacations.md ...",
  },
  {
    "target": "stderr",
    "text": "
  parts/destinations.md",
  },
  {
    "target": "stderr",
    "text": "
  parts/prices.md",
  },
  {
    "target": "stderr",
    "text": "
  parts/destinations.md#Amsterdam",
  },
  {
    "target": "stderr",
    "text": "
  parts/code.md",
  },
  {
    "target": "stderr",
    "text": "
  done.
",
  },
]
`;

exports[`FileInjector > processFile 'fixtures/vacations/vacations.md' {} 1`] = `
[
  [
    "fixtures/vacations/vacations.md",
    "# Vacations

## Destinations

<!--- @@inject: parts/destinations.md --->

## Amsterdam

Amsterdam is full of beautiful canals.

## Orlando

There are many amusement parks in Orlando.

## New York City

It is a wonderful place to catch a show.

## San Francisco

Do not miss out on the Golden Gate Bridge.

<!--- @@inject-end: parts/destinations.md --->

-   Prices
    <!--- @@inject: parts/prices.md --->
    ## Data
    | Name             | City      | Cost   |
    | ---------------- | --------- | ------ |
    | Parade of Lights | Amsterdam | €100/p |
    | Disney World     | Orlando   | $200/p |
    <!--- @@inject-end: parts/prices.md --->

# Highlight Destination (Amsterdam)

<!--- @@inject: parts/destinations.md#Amsterdam --->

## Amsterdam

Amsterdam is full of beautiful canals.

<!--- @@inject-end: parts/destinations.md#Amsterdam --->

## Code

<!--- @@inject-start: parts/code.md --->

## Example

\`\`\`ts
import * as fs from 'fs/promises';
import * as path from 'path';
\`\`\`

<!--- @@inject-end: parts/code.md --->

<pre>
<!--- @@inject: parts/destinations.md -->
</pre>

\`\`\`markdown
<!--- @@inject: parts/destinations.md -->
\`\`\`
",
    "utf8",
  ],
]
`;

exports[`FileInjector > processFile 'fixtures/vacations/vacations.md' {} 2`] = `
[
  {
    "target": "stderr",
    "text": "fixtures/vacations/vacations.md ...",
  },
  {
    "target": "stderr",
    "text": " done.
",
  },
]
`;

exports[`FileInjector > processFile 'vacations.md' { cwd: 'fixtures/vacations/', …(1) } 1`] = `
[
  [
    "_out_/vacations.md",
    "# Vacations

## Destinations

<!--- @@inject: parts/destinations.md --->

## Amsterdam

Amsterdam is full of beautiful canals.

## Orlando

There are many amusement parks in Orlando.

## New York City

It is a wonderful place to catch a show.

## San Francisco

Do not miss out on the Golden Gate Bridge.

<!--- @@inject-end: parts/destinations.md --->

-   Prices
    <!--- @@inject: parts/prices.md --->
    ## Data
    | Name             | City      | Cost   |
    | ---------------- | --------- | ------ |
    | Parade of Lights | Amsterdam | €100/p |
    | Disney World     | Orlando   | $200/p |
    <!--- @@inject-end: parts/prices.md --->

# Highlight Destination (Amsterdam)

<!--- @@inject: parts/destinations.md#Amsterdam --->

## Amsterdam

Amsterdam is full of beautiful canals.

<!--- @@inject-end: parts/destinations.md#Amsterdam --->

## Code

<!--- @@inject-start: parts/code.md --->

## Example

\`\`\`ts
import * as fs from 'fs/promises';
import * as path from 'path';
\`\`\`

<!--- @@inject-end: parts/code.md --->

<pre>
<!--- @@inject: parts/destinations.md -->
</pre>

\`\`\`markdown
<!--- @@inject: parts/destinations.md -->
\`\`\`
",
    "utf8",
  ],
]
`;

exports[`FileInjector > processFile 'vacations.md' { cwd: 'fixtures/vacations/', …(1) } 2`] = `
[
  {
    "target": "stderr",
    "text": "vacations.md ...",
  },
  {
    "target": "stderr",
    "text": " done.
",
  },
]
`;

exports[`FileInjector > processFile('fixtures/vacations/vacations.md', { outputDir: '_out_' }) 1`] = `
[
  [
    "_out_/fixtures/vacations/vacations.md",
    "# Vacations

## Destinations

<!--- @@inject: parts/destinations.md --->

## Amsterdam

Amsterdam is full of beautiful canals.

## Orlando

There are many amusement parks in Orlando.

## New York City

It is a wonderful place to catch a show.

## San Francisco

Do not miss out on the Golden Gate Bridge.

<!--- @@inject-end: parts/destinations.md --->

-   Prices
    <!--- @@inject: parts/prices.md --->
    ## Data
    | Name             | City      | Cost   |
    | ---------------- | --------- | ------ |
    | Parade of Lights | Amsterdam | €100/p |
    | Disney World     | Orlando   | $200/p |
    <!--- @@inject-end: parts/prices.md --->

# Highlight Destination (Amsterdam)

<!--- @@inject: parts/destinations.md#Amsterdam --->

## Amsterdam

Amsterdam is full of beautiful canals.

<!--- @@inject-end: parts/destinations.md#Amsterdam --->

## Code

<!--- @@inject-start: parts/code.md --->

## Example

\`\`\`ts
import * as fs from 'fs/promises';
import * as path from 'path';
\`\`\`

<!--- @@inject-end: parts/code.md --->

<pre>
<!--- @@inject: parts/destinations.md -->
</pre>

\`\`\`markdown
<!--- @@inject: parts/destinations.md -->
\`\`\`
",
    "utf8",
  ],
]
`;

exports[`FileInjector > processFile('fixtures/vacations/vacations.md', { outputDir: '_out_' }) 2`] = `
[
  {
    "target": "stderr",
    "text": "fixtures/vacations/vacations.md ...",
  },
  {
    "target": "stderr",
    "text": " done.
",
  },
]
`;

exports[`FileInjector > processFile('fixtures/vacations/vacations.md', { outputDir: '_out_', dryRun: true }) 1`] = `[]`;

exports[`FileInjector > processFile('fixtures/vacations/vacations.md', { outputDir: '_out_', dryRun: true }) 2`] = `
[
  {
    "target": "stderr",
    "text": "fixtures/vacations/vacations.md ...",
  },
  {
    "target": "stderr",
    "text": " done.
",
  },
]
`;
